### PR TITLE
Support Managed Iceberg streaming writes

### PR DESCRIPTION
**Pull request by ahmedabu98**
Apply some windowing and add a new `triggering_frequency_seconds` parameter to support streaming writes to Iceberg tables.

The triggering frequency controls how often we commit data and create new snapshots

### CODE REVIEW

**Comment by ahmedabu98** [comment]
assign set of reviewers

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergIO.java [hunk 1]
```diff
@@ -77,6 +90,21 @@ public WriteRows to(DynamicDestinations destinations) {
       return toBuilder().setDynamicDestinations(destinations).build();
     }
 
+    /**
+     * Sets the frequency at which data is committed and a new {@link org.apache.iceberg.Snapshot}
+     * is produced.
+     *
+     * <p>Every triggeringFrequency duration, all accumulated {@link
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergIO.java:97]: Might want to soften the language so users know it is approximate, or we'll get complaints about it not lining up perfectly.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java [hunk 2]
```diff
@@ -17,13 +17,21 @@
...
 package org.apache.beam.sdk.io.iceberg;
 
+import static org.apache.beam.sdk.io.iceberg.IcebergWriteSchemaTransformProvider.Configuration;
+
 import com.google.auto.service.AutoService;
+import com.google.auto.value.AutoValue;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:27]: Use checkerframework `Nullable` annotation (unless underlying library like Avro requires you to use theirs). I know it seems like an annotation wouldn't matter, but they have different capabilities in some cases so we standardize on the one that works everywhere it needs to.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java [hunk 3]
```diff
@@ -57,8 +66,59 @@ public String description() {
...
+  public abstract static class Configuration {
+    public static Builder builder() {
+      return new AutoValue_IcebergWriteSchemaTransformProvider_Configuration.Builder();
+    }
+
+    @SchemaFieldDescription("Identifier of the Iceberg table.")
+    public abstract String getTable();
+
+    @SchemaFieldDescription("Name of the catalog containing the table.")
+    @Nullable
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:80]: Make the annotation modify the type, not the method, like so:

`public abstract @Nullable String`

(you will need to switch to the checkerframework annotation for this)

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java [hunk 4]
```diff
@@ -57,8 +66,59 @@ public String description() {
...
+    public abstract String getCatalogName();
+
+    @SchemaFieldDescription("Properties used to set up the Iceberg catalog.")
+    @Nullable
+    public abstract Map<String, String> getCatalogProperties();
+
+    @SchemaFieldDescription("Properties passed to the Hadoop Configuration.")
+    @Nullable
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java:88]: Same throughout

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java [hunk 5]
```diff
@@ -108,11 +119,40 @@ public KV<ShardedKey<Row>, Row> apply(Row elem) {
                 "Write remaining rows to files",
                 new WriteGroupedRowsToFiles(catalogConfig, dynamicDestinations));
 
+    PCollection<FileWriteResult> writeUngroupedResultPColl = writeUngroupedResult.getWrittenFiles();
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:122]: Validate here also: if the input is unbounded then triggering frequency is set. I know you check it above, but also do it here so the local code is validated.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java [hunk 6]
```diff
@@ -108,11 +119,40 @@ public KV<ShardedKey<Row>, Row> apply(Row elem) {
...
+
+    if (triggeringFrequency != null) {
+      // for streaming pipelines, re-window both outputs to keep Flatten happy
+      writeGroupedResult =
+          writeGroupedResult.apply(
+              "RewindowGroupedRecords",
+              Window.<FileWriteResult>into(new GlobalWindows())
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:129]: I think you also need to set discarding mode in both of these re-windowings, too. Or validate that they are already set up that way.

File: sdks/java/io/iceberg/src/test/java/org/apache/beam/sdk/io/iceberg/IcebergIOIT.java [hunk 7]
```diff
@@ -307,4 +314,38 @@ public void testWritePartitionedData() {
     assertThat(
         returnedRecords, containsInAnyOrder(INPUT_ROWS.stream().map(RECORD_FUNC::apply).toArray()));
   }
+
+  @Test
+  public void testStreamingWrite() {
```
Comment (kennknowles) [inline sdks/java/io/iceberg/src/test/java/org/apache/beam/sdk/io/iceberg/IcebergIOIT.java:319]: I suggest a couple more tests where the user has set up their PCollection differently, like if it started out with accumulating mode, or if they set a weird trigger in the middle of their pipeline.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergIO.java (hunk 1 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergIO.java]: good call, I softened it up

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java (hunk 2 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java]: Done

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java (hunk 3 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java]: Thanks for this. AutoValue wasn't recognizing checkerframework's `Nullable` annotation as it is placed here, so I switched to javax. Looks like I just had the annotation in the wrong place.

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java (hunk 5 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:122]: Thanks, setting discarding mode

Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:122]: Done

File: sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java (hunk 6 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/WriteToDestinations.java:129]: Thanks, setting discarding mode

File: sdks/java/io/iceberg/src/test/java/org/apache/beam/sdk/io/iceberg/IcebergIOIT.java (hunk 7 above)
Comment (ahmedabu98) [reply sdks/java/io/iceberg/src/test/java/org/apache/beam/sdk/io/iceberg/IcebergIOIT.java:321]: Added a test with fixed windows and accumulating mode. Let me know if there's anything particular we should test for

**Comment by ahmedabu98** [comment]
Thanks @kennknowles, this is ready for another look

**Review by kennknowles — APPROVED** [review]

**Pull request merged**